### PR TITLE
BUG: fix str.replace('.','') should replace every character

### LINKED ISSUE #24804: str.replace('.','') should replace every character?
#### Code Sample (pandas 0.23.0)

In [1]:
```
import pandas as pd
s = pd.Series(['abc','123'])
s.str.replace('.','',regex = True)
```
Out [1]:
```
0    abc
1    123
dtype: object
```

#### Problem description

Hi everyone, I was showing str.replace to a colleague and how it uses regex by default and when I entered pd.Series.str.replace('.','') I expected every character to be removed but instead nothing happened. Is that a bug or is there something I don't understand?

#### Expected Output: empty strings ""
Out [1]:
```
0    
1    
dtype: object
```

#### Output of pandas 0.23.0

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.0
pytest: 3.5.1
pip: 18.0
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.1.1
sphinx: 1.7.4
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.3
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: 2.4.11
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.4
lxml: 4.2.3
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.1.18
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #24804
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
there are no tests for the deprecation

File: pandas/core/reshape/melt.py [hunk 1]
```diff
@@ -420,7 +420,8 @@ def melt_stub(df, stub, i, j, value_vars, sep):
         newdf = melt(df, id_vars=i, value_vars=value_vars,
                      value_name=stub.rstrip(sep), var_name=j)
         newdf[j] = Categorical(newdf[j])
-        newdf[j] = newdf[j].str.replace(re.escape(stub + sep), "")
+        newdf[j] = newdf[j].str.replace(re.escape(stub + sep), "",
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/core/reshape/melt.py]: Thanks very much for taking your time to reivew @jreback . Yes, there is test in tests/reshape/test_melt.py that will fail, because previously the default is True, and now is None. And apparently, in this melt, we want a True for regex.

File: pandas/core/strings.py [hunk 2]
```diff
@@ -456,9 +456,10 @@ def str_replace(arr, pat, repl, n=-1, case=None, flags=0, regex=True):
     flags : int, default 0 (no flags)
         - re module flags, e.g. re.IGNORECASE
         - Cannot be set if `pat` is a compiled regex
-    regex : boolean, default True
+    regex : boolean, default None
         - If True, assumes the passed-in pattern is a regular expression.
         - If False, treats the pattern as a literal string
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/core/strings.py]: Yes, I understand that @jreback it's a quite disruptive change. This change is following the proposal from @TomAugspurger in https://github.com/pandas-dev/pandas/pull/24809#discussion_r248545924 . The purpose is to make this argument more explicit.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/strings.py]: >  I would be ok with forcing the passing of regex always, possibly in preparation of changing the default to regex=False

@jreback can you clarify what you mean by "forcing"? If we raise when `regex` is not specified, that would be an API breaking change.

My proposal is the preserve the previous behavior, but warn when a length-1 regex is detected. Then we can get to the documented behavior of `regex=True`.

File: pandas/core/strings.py [hunk 3]
```diff
@@ -456,9 +456,10 @@ def str_replace(arr, pat, repl, n=-1, case=None, flags=0, regex=True):
...
+        - If pat is single special character, default regex is False.
```
Comment (TomAugspurger) [inline pandas/core/strings.py]: This should mention something about a warning being issued when regex is not specified. Something like

    - If `pat` is a single character and `regex` is not specified, `pat` is interpreted as a
       string literal. If `pat` is also a regular expression symbol, a warning is issued that
       in the future `pat` will be interpreted as a regex, rather than a literal.

File: pandas/core/strings.py [hunk 4]
```diff
@@ -577,6 +578,12 @@ def str_replace(arr, pat, repl, n=-1, case=None, flags=0, regex=True):
         if callable(repl):
             raise ValueError("Cannot use a callable replacement when "
                              "regex=False")
+        # if regex is default None, and a single special character is given
+        # in pat, still take it as a literal, and raise the Future warning
+        if regex is None and len(pat) == 1 and re.findall(r"\W", pat):
```
Comment (TomAugspurger) [inline pandas/core/strings.py]: I don't understand the `'\W'` here. I think that's not quite right for non-word characters that are not regex symbols. Something like `@`

Does python have a list of regex symbols?

Is the heuristic `re.escape(pat) == pat` good enough? 

```python
In [46]: re.escape('@')
Out[46]: '@'

In [47]: re.escape('^')
Out[47]: '\\^'

```

Comment (charlesdong1991) [reply pandas/core/strings.py]: @TomAugspurger oh, indeed it's wrong! nice catch!! thanks for the correction!!

File: doc/source/whatsnew/v0.25.0.rst [hunk 5]
```diff
@@ -40,9 +40,7 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
--
--
--
+- :func:`Series.str.replace`, the default regex for single character might be deprecated.
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.0.rst]: This will need to be updated.

File: pandas/core/strings.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/core/strings.py]: sure! @TomAugspurger

Comment (charlesdong1991) [reply pandas/core/strings.py]: ehh, @TomAugspurger  escape somehow changed... wired
i did get same output the first time i tested on my jupyter notebook:
re.escape("@") -> "@"
but now, i cannot replicate this and get re.escape("@") -> "\\@" ...
so i will go for this option:
pat in list("[\^$.|?*+()]")

**Comment by charlesdong1991** [comment]
any follow-up code change requests and reviews? @jreback @TomAugspurger

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -42,6 +42,7 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
+- :func:`Series.str.replace`, when pat is single special regex character and regex is not defined, regex is by default False for now, but this might be deprecated in the future.
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: needs an issue reference; use double backticks on ``False``, use single backticks on keywordds.

**Review by jreback — CHANGES_REQUESTED** [review]
I would rather simply go whole hog and just change ``regex=None`` as a warning for *all*, meaning you must change to an explicit True/False. It makes it more clear. But this would then be a rather large change.

File: doc/source/whatsnew/v0.25.0.rst (hunk 6 above)
Comment (charlesdong1991) [reply doc/source/whatsnew/v0.25.0.rst]: thanks changed!

**Comment by charlesdong1991** [comment]
yes, indeed will be a kind of api change. shall I open another issue to make this change if other maintainers agree with this? Or do you want to do this change in this PR? @jreback

**Comment by WillAyd** [comment]
So as an alternate what if we simply detected the case where the pattern provided is a single / special regex character with `regex=True` and raise a FutureWarning when that happens, noting that we will change the behavior to evaluate that as a regular expression regardless in the future

**Comment by charlesdong1991** [comment]
@WillAyd based on your description, i think it's kind of what this PR is doing, right?

**Comment by WillAyd** [comment]
I'm saying don't even update the signature of the function, simply do something like:

```python
if regex=True and pater in re_char_blacklist:
    raise FutureWarning("Single character patterns will be evaluated as regexes in the future")
```

Or something to the effect. Would minimize a lot of the volatility going on here

**Comment by TomAugspurger** [comment]
@WillAyd the `raise` in your pseudocode should probably be a `warnings.warn`?

I (think?) then there is no way to get the desired future behavior without warning. IIUC, your prosposal would warn on

```python
pd.Series(['aa']).str.replace('.', 'b', regex=True)
```

I think we'll need a period of

1. `regex=None` (default): warning and output `aa`
2. `regex=True`: no warning, output `bb`
3. `regex=False`: no warning, output `aa`

**Comment by WillAyd** [comment]
Yea sorry meant warn instead of raise. Why do we need regex=None at all? I was hoping to avoid that API change

**Comment by TomAugspurger** [comment]
Does the rest of 
https://github.com/pandas-dev/pandas/pull/24935#issuecomment-468353395 answer your question? The short version is (IIUC), your proposal wouldn't allow for option 2 (`regex=True` interprets `.` as a regex, and doest *not* warn).

**Comment by WillAyd** [comment]
Hmm unless I'm mistaken my proposal would only affect option two. So something like:

```python
if regex=True and pater in {'.', '^', '$', '*', '+', '?', '}'}:
   warnings.warn("You've passed a single special character in with regex=True which is evaluated as a single character. This will change in the future to match regex behavior")
```

Warns the user about this behavior in the future but doesn't impact their current code, allowing for a smoother deprecation in the future.

To your point though the result of option 2 would still stay 'aa' as is today just with the warning

**Comment by TomAugspurger** [comment]
Right, so I think we understand each other.

My argument is that we shouldn't add a warning without a concrete action a
user can take to avoid it. Else it'll just annoy them :)

**Comment by WillAyd** [comment]
Ah I gotcha. So in my mind if a user is passing a single special character and regex=True (whether explicitly or via the default argument) then it is a current "misuse" of that parameter. 

If the warning said something like "explicitly pass regex=False to maintain current behavior" would that be better?

**Comment by TomAugspurger** [comment]
Then we'd need to wait to actually fix the bug. Since we have a PR now, I'd
prefer to go with the (pretty-well established) pattern of using None to
indicate
that the user needs to take explicit action to avoid the warning.

**Comment by charlesdong1991** [comment]
Many thanks for your feedback and discussions @TomAugspurger @WillAyd 
My English is not very good, and i might get distracted by your discussions, could you give me any hint on what kind of change I need to make (or give a general direction to make me a bit clearer)? as far as i understand from your discussions, seems warning messages should be changed a bit? @WillAyd @TomAugspurger  many thanks and sorry for asking this.

**Comment by TomAugspurger** [comment]
No worries. I'd recommend waiting, I don't think a decision has been made
yet.

File: doc/source/whatsnew/v0.25.0.rst [hunk 7]
```diff
@@ -43,6 +43,7 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
+- :func:`Series.str.replace`, when pat is single special regex character and regex is not defined, regex is by default ``False`` for now, but this might be deprecated in the future. (:issue:`24804`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use double back-ticks around ``pat``.

..is single special regex character'.. (such as ``.|\`` etc).

File: pandas/core/strings.py [hunk 8]
```diff
@@ -456,9 +456,13 @@ def str_replace(arr, pat, repl, n=-1, case=None, flags=0, regex=True):
     flags : int, default 0 (no flags)
         - re module flags, e.g. re.IGNORECASE
         - Cannot be set if `pat` is a compiled regex
-    regex : bool, default True
+    regex : boolean, default None
```
Comment (jreback) [inline pandas/core/strings.py]: default is still True

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/strings.py (hunk 8 above)
Comment (charlesdong1991) [reply pandas/core/strings.py]: so, you mean we now dont raise a warning anymore when char is a special single character while user doesn't explicitly specify `regex`?
if set default to `True`, and i use @TomAugspurger example:
```python
pd.Series(['aa']).str.replace('.', 'b')
```
we will get:
1. `default` or explicitly `regex=True`, output is `bb`, no warning
2. `regex=False`, output is `aa`, no warning.
is it correct? @WillAyd @TomAugspurger @jreback

**Comment by TomAugspurger** [comment]
@WillAyd can you summarize your aversion to the (typical) use of None as a backwards compatible sentinel value for when we should warn?

**Comment by WillAyd** [comment]
Not terribly against it if that’s our process

**Comment by jreback** [comment]
can you merge master

**Comment by charlesdong1991** [comment]
Hi, @jreback I just merged master and fix conflict!

**Comment by jreback** [comment]
something still failing

**Comment by charlesdong1991** [comment]
@jreback i just merged the master and fix test error

**Comment by WillAyd** [comment]
This one has been around for a while. Just so we are all on the same page - is the intention here definitely to force users down the road to explicitly provide regex=True or regex=False? I know that was mentioned in comments above just not sure we agreed to it.

An easy alternative would just be to raise a DeprecationWarning when a single special character gets passed and actually change behavior at a later date, which I think is what @TomAugspurger suggests in https://github.com/pandas-dev/pandas/issues/24804#issuecomment-457315905

**Comment by TomAugspurger** [comment]
I don’t see a reason to make the regex flag required.

**Comment by jreback** [comment]
ok, re-reading: https://github.com/pandas-dev/pandas/pull/24809#discussion_r248670917

that looks like a good idea, not sure where that is in implementation.

**Comment by charlesdong1991** [comment]
@jreback thanks for your comment! could you explain a little bit more? I think the change I made in this PR was following @TomAugspurger suggestion in 24809

**Comment by jreback** [comment]
@charlesdong1991 can you merge master and we'll see where this is

**Comment by charlesdong1991** [comment]
@jreback I merged to master and resolved conflicts, pls take a look

**Comment by TomAugspurger** [comment]
I just reread through the issue, and my thinking has changed a bit.

Do we think that `.str.replace('.', '...')` replacing every character is actually the most *useful* thing for users? Or is the change primarily motivated by the inconsistency?

If we think that most users will expect `.str.replace(".")` to only replace periods, then perhaps we just document that we don't auto-regex single-character `pat` (`.`, `^`, `$`, `|`) and tell users to pass `re.compile(pat)` if they want that?

**Comment by WillAyd** [comment]
> Or is the change primarily motivated by the inconsistency

Definitely inconsistency IMO. I don't see that particular operation being useful but I don't think we should be special casing either, as our general handling of regexes in str replacement ops is all over the place

**Comment by TomAugspurger** [comment]
What other inconsistencies do you see?

FWIW, I perfectly fine with documenting that length-1 `pat` aren't
auto-converted to regex, even when `regex` is true.

I think the state I want to avoid is surprising future users with `.`
replacing every character, just because they aren't familiar with regex.

So either we document that length-1 characters are special cased, or
deprecate `regex=True`. I don't have a good feeling for which is better.

**Comment by WillAyd** [comment]
I wouldn't be opposed to really doing nothing on this atm. The behavior can definitely be surprising as noted by the OP of the original issue, but to your point I don't see why the strictly standard behavior is useful either.

May be worth just punting this for a more comprehensive alignment of the regex keywords across string ops if it makes sense to tackle then

**Comment by TomAugspurger** [comment]
By "doing nothing" do you include "document the current behavior of special
casing length-1 patterns"? If so, then I agree that's best for now :)

**Comment by WillAyd** [comment]
Agreed - @charlesdong1991 make sense to you?

**Comment by charlesdong1991** [comment]
ok, it does.
